File: lee/src/main/java/org/example/lee/题目/动态规划/a_买卖股票的最佳时机.java
package org.example.lee.题目.动态规划;

public class a_买卖股票的最佳时机 {

  /**
   * max profit dp[i]
   *
   * @param prices prices
   * @return int
   */
  public int maxProfit(int[] prices) {
    if (prices.length <= 1) {
      return 0;
    }
    int min = prices[0], key = prices[1] - prices[0];
    for (int i = 1; i < prices.length; i++) {
      min = Math.min(min, prices[i]);
      key = Math.max(key, prices[i] - min);
    }
    return key;
  }
}

File: lee/src/main/java/org/example/lee/题目/贪心/a_买卖股票的最佳时机.go
package 贪心

import "math"

func maxProfit(prices []int) int {
	min1, max1 := math.MaxInt32, 0
	for _, v := range prices {
		min1 = min(min1, v)
		max1 = max(max1, v-min1)
	}
	return max1
}
